### ISSUE TITLE
[BUG]  Distorted image with "hires 0"

### ISSUE THREAD

**Issue by vanfanel** [Russian-Doom/russian-doom#282]
### Russian Doom version

Russian Heretic 5.1 GIT

### What Operating System are you using?

Linux x86_64

### If Other OS, please describe

Ubuntu

### Relevant hardware info

Intel GPU

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

Hi!

On Crispy Doom, I can simply change:

`#define hires 1`

to 

`#define hires 0`

in `src/i_video.h`, and that way I can do non-hires builds that look smooth (and use almost no CPU!).
However, in Russian Heretic, doing a `#define hires 0` build results on a distorted image.
I am almost sure it's something trivial... could it be fixed?

Thanks!

### Steps to reproduce the behaviour.

Explain how to reproduce
1. Build Russian Heretic with `#define hires 0`
2. Launch the game

### Your configuration

english_language              1   
autoload_global_pwad1         ""  
autoload_global_pwad2         ""  
autoload_global_pwad3         ""  
autoload_global_pwad4         ""  
autoload_registered_pwad1     ""  
autoload_registered_pwad2     ""  
autoload_registered_pwad3     ""  
autoload_registered_pwad4     ""  
autoload_retail_pwad1         ""  
autoload_retail_pwad2         ""  
autoload_retail_pwad3         ""  
autoload_retail_pwad4         ""  
use_mouse                     1   
use_controller                1   
controller_invert_move        0   
controller_invert_strafe      0   
controller_invert_turn        0   
controller_invert_vlook       1   
controller_bind_move          1   
controller_bind_strafe        0   
controller_bind_turn          2   
controller_bind_vlook         -1  
snd_musicdevice               5   
snd_sfxdevice                 3   
chatmacro0                    "(null)"
chatmacro1                    "(null)"
chatmacro2                    "(null)"
chatmacro3                    "(null)"
chatmacro4                    "(null)"
chatmacro5                    "(null)"
chatmacro6                    "(null)"
chatmacro7                    "(null)"
chatmacro8                    "(null)"
chatmacro9                    "(null)"
fullscreen                    1
aspect_ratio                  2
video_display                 0
vga_porch_flash               0
integer_scaling               0
startup_delay                 35
resize_delay                  70
window_title_short            1
window_width                  640
window_height                 361
window_border                 1
fullscreen_width              0
fullscreen_height             0
grabmouse                     1
mouse_acceleration            2.000000
mouse_threshold               10
snd_samplerate                44100
snd_cachesize                 67108864
snd_maxslicetime_ms           28
snd_musiccmd                  ""
snd_dmxoption                 "-opl3"
opl_io_port                   0x388
show_endoom                   0
vanilla_keyboard_mapping      1
video_driver                  ""
window_position_x             500
window_position_y             281
player_name                   "pi"
libsamplerate_scale           0.650000
fluidsynth_sf_path            ""
timidity_cfg_path             ""
gus_patch_path                "gus_patches/"
gus_ram_kb                    1024
vsync                         1
aspect_ratio_correct          1
uncapped_fps                  1
show_fps                      0
smoothing                     1
force_software_renderer       0
png_screenshots               1
flashing_hom                  0
screenblocks                  10
level_brightness              0
detaillevel                   0
brightness                    1.000000
usegamma                      8
color_saturation              1.000000
show_palette                  1
r_color_factor                1.000000
g_color_factor                1.000000
b_color_factor                1.000000
local_time                    0
show_messages                 1
messages_alignment            0
messages_timeout              4
message_fade                  1
message_pickup_color          0
message_secret_color          15
message_system_color          0
message_chat_color            5
draw_shadowed_text            1
automap_stats                 1
automap_level_time            1
automap_total_time            0
automap_coords                0
automap_overlay               0

### Provide a Log

                                Russian Heretic 5.1                               
M_LoadDefaults: Загрузка системных стандартов.
Сохранение файла конфигурации:
        russian-heretic.ini
V_Init: allocate screens.
Z_Init: Init zone memory allocation daemon.
zone memory: 0x7f9006e010, 20 MB allocated for zone
W_Init: Init WADfiles.
 adding: heretic.wad
 adding: heretic_music.wad
Savegames folder:
        savegames/
I_SDLMusic: Using GUS patch set from:
        gus_patches/
NET_Init: Init network subsystem.
MN_Init: Init menu system.
R_Init: Init Heretic refresh daemon......[........                           ]
P_Init: Init Playloop state.
I_Init: Setting up machine state.
I_InitController: X360 Controller
S_Init: Setting up sound.
I_SDL_PrecacheSounds: Precaching all sound effects - [........................]
D_CheckNetGame: Checking network game status.
SB_Init: Loading patches.
Starting game: "Heretic".

**Comment by JNechaevsky** [comment]
That's true. I'm afraid `hires` macro is not supposed to changed, since many functions like visplanes rendering are now `hires 1`-friendly only. This is also why it's still macrocized and not allowed to be changed via config-file.

I'm not entirely sure how to make functions `hires 0` friendly (and how to handle "Detail level" then!), but you still can set "Detail level" to low, which will emulate standard 320x200 screen and will draw less columns.

**Comment by vanfanel** [comment]
@JNechaevsky Ah, I see. But using "Detail Level" to low, tittle screen and menus will still look "grainy".

**Comment by JNechaevsky** [comment]
But how them are supposed to look? There no HQ smooth upscaler, they are always 320x200 (well, 560x200 for widescreen assets) pixels upscaled to screen size. Drawing without 2x scaling is possible and used for crosshair and palette indicator:

![image](https://user-images.githubusercontent.com/21193394/137631383-6e0908ff-db68-45ae-b4f2-61ef08466d41.png)

Such unscaled drawing will probably also be broken in `hires 0`.

**Comment by vanfanel** [comment]
@JNechaevsky There is bilinear filtering that SDL2 does "for free" (using OpenGL under the hood).
I always change "nearest" for "linear" here:

https://github.com/JNechaevsky/russian-doom/blob/2601e283cbf4b731d676fd665be9e1892c41f0d5/src/i_video.c#L1532

So things with `hires 0` and this change look a bit blurry, as in a CRT monitor, like those games were designed to look.

This technique of mine doesn't work if `hires 0` is not possible.

**Comment by JNechaevsky** [comment]
Wait, there is already an option for SDL's pixel smoothing in the port. It is right here:

**Comment by JNechaevsky** [comment]
![image](https://user-images.githubusercontent.com/21193394/137642458-402ee20b-7719-4548-b930-d2683f0f85cc.png)

**Comment by vanfanel** [comment]
@JNechaevsky That doesn't do anything I can see. Does it make any difference for you?

**Comment by JNechaevsky** [comment]
Sure, it's working for me, otherwise it wouldn't be there. Could you please double-check - will be there any effect after program restart? Not sure what is happing on your side: either how-swapping is not working, or smooth scaling does not working at all.

**Comment by fabiangreffrath** [comment]
It's possible to break this feature by mis-adjusting the max_scaling_buffer_pixels config key value.

**Comment by vanfanel** [comment]
@JNechaevsky Yes, no difference here. I tried to exit the game and run it again, too.

**Comment by JNechaevsky** [comment]
Ah. Does it help if you'll comment out https://github.com/JNechaevsky/russian-doom/blob/master/src/i_video.c#L1305-L1309 like:
```
{
//  SDL_SetHintWithPriority(SDL_HINT_RENDER_DRIVER, force_software_renderer ?
//                          "software" : "opengles2", SDL_HINT_OVERRIDE);
}
```

or replace `opengles2` with `direct3d`?

> It's possible to break this feature by mis-adjusting the max_scaling_buffer_pixels config key value.

Looks likes I'm still using old code, there is no `max_scaling_buffer_pixels` config variable in RD.

**Comment by vanfanel** [comment]
@JNechaevsky I can't do that change because I only have OpenGL(ES) here, this is not a Windows machine and I don't have any Windows machines to build on them :(

**Comment by JNechaevsky** [comment]
No-no, it's allright. If you'll comment out that lines, SDL itself will try to choose render. You could also replace to `"opengles2"` with `"opengl"`, but from my `-timedemo` benchmarks, OpenGL giving less performance than standard Direct3D on Windows, and even less than OpenGles2.

**Comment by vanfanel** [comment]
Removing the SDL_SetHint() doesn't make any difference to the image either.

**Comment by JNechaevsky** [comment]
I'm out of ideas then. Is it an option to buy a Windows PC? 😋

**Comment by vanfanel** [comment]
@JNechaevsky Not really, I am a die-hard GNU/Linux fan. The problem wouldn't be the PC, but the OS! XD

Don't worry, anyway russian-heretic looks and plays absolutely fantastic!

**Comment by JNechaevsky** [comment]
@vanfanel, recently I have ported all rendering optimizations into Hexen, as well as low graphics detail mode (toggling available only in Display Options, `F5` is still used for player suicide). Low detail should do it's job just fine, it's not just a post-processing effect.

There is still potential for some small performance tuning, but in practice this is not really needed, since Hexen is still not able to load maps with extended nodes _(read: huge levels)_, and on vanilla levels these optimizations just making a minor performance improvements on low-end devices.

I've also reverted OpenGles to system-preffered. Despite of working faster, Gles is not very friendly with various OSD widgets and some screen-capturing software on Windows OS.

There are also a lot of improvements, and as always, no proper changelog composed yet. 🤤

As about `hires` macro, well, no, better leave it as is, it's not supposed to be changed by user anyways, and making it variable for toggling between 320x200 / 160x200 will bring a lot of unnecessary headache.

Looks likes we are done here?
